docs/design/simplify-custom-resource-metrics-api.md

```diff
@@ -0,0 +1,198 @@
+| path: [status, sub] \n valueFrom: [ready] |  x.status.sub.map(y, x.status.sub[y].ready) |
+| commonLabels: \n custom_metric: "yes" | [{ "custom_metric":"yes" }] |
+| labelsFromPath: "*": [metadata, labels] | [x.metadata.labels] |
+| labelsFromPath \n foo: [metadata, labels, foo] | [{'name': x.metadata.name}] |
+
+## Example
+### CR
+```
+kind: Foo
+apiVersion: myteam.io/vl
+metadata:
+    annotations:
+        bar: baz
+        qux: quxx
+    labels:
+        foo: bar
+    name: foo
+spec:
+    version: v1.2.3
+    order:
+        - id: 1
+          value: true
+        - id: 3
+          value: false
+    replicas: 1
+status:
+    phase: Pending
+    active:
+        type-a: 1
+        type-b: 3
+    conditions:
+        - name: a
+          value: 45
+        - name: b
+          value: 66
+    sub:
+        type-a:
+            active: 1
+            ready: 2
+        type-b:
+            active: 3
+            ready: 4
+    uptime: 43.21
+```
+
+### Existing API - CustomResourceStateMetrics
+```
+kind: CustomResourceStateMetrics
+spec:
+  resources:
+    - groupVersionKind:
+        group: myteam.io
+        kind: "Foo"
+        version: "v1"
+      # labels can be added to all metrics from a resource
+      commonLabels:
+        crd_type: "foo"
+      labelsFromPath:
+        name: [metadata, name]
+      metrics:
+        - name: "ready_count"
+          help: "Number Foo Bars ready"
+          each:
+            type: Gauge
+            gauge:
+              # targeting an object or array will produce a metric for each element
+              # labelsFromPath and value are relative to this path
+              path: [status, sub]
+
+              # if path targets an object, the object key will be used as label value
+              # This is not supported for StateSet type as all values will be truthy, which is redundant.
+              labelFromKey: type
+              # label values can be resolved specific to this path
+              labelsFromPath:
+                active: [active]
+              # The actual field to use as metric value. Should be a number, boolean or RFC3339 timestamp string.
+              valueFrom: [ready]
+          commonLabels:
+            custom_metric: "yes"
+          labelsFromPath:
+            # whole objects may be copied into labels by prefixing with "*"
+            # *anything will be copied into labels, with the highest sorted * strings first
+            "*": [metadata, labels]
+            "**": [metadata, annotations]
+
+            # or specific fields may be copied. these fields will always override values from *s
+            name: [metadata, name]
+            foo: [metadata, labels, foo]
+```
+
+### Proposed API - CustomResourceStateMetricsV2
+```
+kind: CustomResourceStateMetricsV2
+spec:
+  resources:
+    - groupVersionKind:
+        group: myteam.io
+        kind: "Foo"
+        version: "v1"
+      mode: for_loop # or merged
+      metrics:
+        - name: "ready_count"
```

Comment:
how is the metrics type defined?

Reply:
I presume we'll still need fields with hardcoded values, eg., `type`.

Reply:
Noting here that there is still the need for non-path-based constructs, like so:
- `type` (used to check OpenMetrics conformance, and if `value` field is applicable)
- `path` (relative paths when used with `labelsFrom*`)
	- non-relative
	- relative
	- multi-relative (relative to different paths) (KSM does not supports this right now)
- `commonLabels`
- [`list`](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/customresourcestate-metrics.md#stateset) (maybe replaced by a CEL-based alternative)

Something we might want to persue under the configuration revamp is to keep such constructs to a minimum (and try to replace most of these using CEL, wherever possible), as we may find ourselves implementing custom logic and introducing complexity as such constructs grow over time.